src/TokensTracker/TrackerBase.cs

```diff
@@ -0,0 +1,146 @@
+            return transferPairs;
+        }
+
+        protected static byte[] Key(byte prefix, ISerializable key)
+        {
+            byte[] buffer = new byte[key.Size + 1];
+            using (MemoryStream ms = new MemoryStream(buffer, true))
+            using (BinaryWriter writer = new BinaryWriter(ms))
+            {
+                writer.Write(prefix);
+                key.Serialize(writer);
+            }
+            return buffer;
+        }
+
+        protected void Put(byte prefix, ISerializable key, ISerializable value)
+        {
+            _levelDbSnapshot.Put(Key(prefix, key), value.ToArray());
+        }
+
+        protected void Delete(byte prefix, ISerializable key)
+        {
+            _levelDbSnapshot.Delete(Key(prefix, key));
+        }
+
+        protected TransferRecord GetTransferRecord(UInt160 asset, Array stateItems)
+        {
+            if (stateItems.Count < 3)
+            {
+                return null;
+            }
+            var fromItem = stateItems[0];
+            var toItem = stateItems[1];
+            var amountItem = stateItems[2];
+            if (fromItem.NotNull() && fromItem is not ByteString)
+                return null;
+            if (toItem.NotNull() && toItem is not ByteString)
+                return null;
+            if (amountItem is not ByteString && amountItem is not Integer)
+                return null;
+
+            byte[] fromBytes = fromItem.IsNull ? null : fromItem.GetSpan().ToArray();
+            if (fromBytes != null && fromBytes.Length != UInt160.Length)
+                return null;
+            byte[] toBytes = toItem.IsNull ? null : toItem.GetSpan().ToArray();
+            if (toBytes != null && toBytes.Length != UInt160.Length)
+                return null;
+            if (fromBytes == null && toBytes == null)
+                return null;
+
+            var from = fromBytes == null ? UInt160.Zero : new UInt160(fromBytes);
+            var to = toBytes == null ? UInt160.Zero : new UInt160(toBytes);
+            if (stateItems.Count == 3)
+            {
+                return new TransferRecord(asset, from, to, null, amountItem.GetInteger());
+            }
+            if (stateItems.Count == 4 && (stateItems[3] is ByteString tokenId))
+            {
+                return new TransferRecord(asset, from, to, tokenId, amountItem.GetInteger());
+            }
+            return null;
+        }
+
+        protected JObject ToJson(TokenTransferKey key, TokenTransfer value, bool isSent)
+        {
+            JObject transfer = new JObject();
+            transfer["timestamp"] = key.TimestampMS;
+            transfer["assethash"] = key.AssetScriptHash.ToString();
+            transfer[isSent ? "toaddress" : "fromaddress"] = value.UserScriptHash == UInt160.Zero ? null : value.UserScriptHash.ToAddress(_neoSystem.Settings.AddressVersion);
```

Comment:
This changes NEP-17 tracker interface, it unconditionally used `transferaddress` field before and I think it should remain the same as long as we have separate `sent` and `received` arrays in the answer.